File: projects/06/assembler/hack_parser.py
from instruction import Instruction
from utils import InstructionType


class HackParser:

    def __init__(self, file) -> None:
        self.filename = file.split('.')[0]
        self.read_file(file)
        self.current_instruction = Instruction()

    @property
    def hack_file(self):
        self._hack_file = open(f'{self.filename}-py.hack', 'x')
        return self._hack_file

    def read_file(self, file) -> None:
        self._asm_file = open(file, "r")
        self._asm_file_lines = len(self._asm_file.readlines())
        self._asm_file.seek(0)
        self._lines_read = 0
        self.instruction_line = -1
    
    def reset_file(self):
        self._asm_file.seek(0)
        self._lines_read = 0
        self.current_instruction = Instruction()

    @property
    def has_more_lines(self) -> bool:
        return self._asm_file_lines > self._lines_read

    def advance(self) -> None:
        current_line = self._asm_file.readline().strip()
        self._lines_read += 1

        self.current_instruction.process_instruction(current_line)
        instruction_type = self.get_instruction_type()
        if instruction_type == InstructionType.A or instruction_type == InstructionType.C:
            self.instruction_line += 1
        
    def get_instruction_type(self) -> InstructionType:
        return self.current_instruction.get_instruction_type()
    
    def get_symbol(self) -> str:
        return self.current_instruction.get_symbol()
        
    def get_dest(self) -> str:
        return self.current_instruction.get_destination()
    
    def get_comp(self) -> str:
        return self.current_instruction.get_computation()
    
    def get_jump(self) -> str:
        return self.current_instruction.get_jump()
       

File: projects/06/go-assembler/hackParser.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type HackParser struct {
	filename           string
	HackFile           *os.File
	asmFile            *os.File
	cachedLine         *string
	InstructionLine    int
	scanner            *bufio.Scanner
	currentInstruction Instruction
}

func NewHackParser(file string) (HackParser, error) {
	newFile, err := os.Create(fmt.Sprintf("%v-go.hack", file[:len(file)-4]))
	if err != nil {
		fmt.Println("ERROR: Failed to create file:", err)
		return HackParser{}, err
	}

	return HackParser{
		filename:           file,
		HackFile:           newFile,
		InstructionLine:    -1,
		currentInstruction: NewEmptyInstruction(),
	}, nil
}

func (h *HackParser) ReadFile() error {
	file, err := os.Open(h.filename)
	if err != nil {
		fmt.Println("ERROR: Failed to open file:", err)
		return err
	}

	h.asmFile = file
	h.scanner = bufio.NewScanner(h.asmFile)
	return nil
}

func (h *HackParser) ResetFile() {
	h.asmFile.Seek(0, 0)
	h.scanner = bufio.NewScanner(h.asmFile)
	h.cachedLine = nil
	h.currentInstruction = NewEmptyInstruction()
}

func (h *HackParser) HasMoreLines() bool {
	if h.cachedLine != nil {
		return true
	}
	if !h.scanner.Scan() {
		return false
	}
	currentLine := h.scanner.Text()
	h.cachedLine = &currentLine
	return true
}

func (h *HackParser) Advance() {
	if h.scanner == nil {
		fmt.Println("Empty scanner, returning...")
		return
	}

	var currentLine string
	if h.cachedLine != nil {
		currentLine = *h.cachedLine
		h.cachedLine = nil
	} else {
		h.scanner.Scan()
		currentLine = h.scanner.Text()
	}

	currentLine = strings.TrimSpace(currentLine)
	h.currentInstruction.ProcessInstruction(currentLine)
	if h.GetInstructionType() == InstructionA || h.GetInstructionType() == InstructionC {
		h.InstructionLine++
	}
}

func (h *HackParser) GetInstructionType() InstructionType {
	return h.currentInstruction.GetInstructionType()
}

func (h *HackParser) GetSymbol() string {
	return h.currentInstruction.GetSymbol()
}

func (h *HackParser) GetDest() string {
	return h.currentInstruction.GetDestination()
}

func (h *HackParser) GetComp() string {
	return h.currentInstruction.GetComputation()
}

func (h *HackParser) GetJump() string {
	return h.currentInstruction.GetJump()
}
